File: packages/totem-sdk/packages/recursive-mast/src/content-keys.ts
/**
 * Content-addressed key scheme for recursive MAST policy storage.
 *
 * Every object is addressed by its cryptographic hash, enabling
 * storage-agnostic retrieval across Hypercore, HTTP, filesystem,
 * or any other content-addressed backend.
 *
 * Key formats:
 *   policy:<policyId>:manifest:<version>
 *   script:<scriptHash>
 *   proof:<policyRoot>:<scriptHash>
 *   bundle:<bundleHash>
 */

import { sha3_256, bytesToHex } from '@totemsdk/core';

export const KEY_PREFIX = {
  POLICY_MANIFEST: 'policy',
  MANIFEST_DIGEST: 'manifest',
  SCRIPT: 'script',
  PROOF: 'proof',
  BUNDLE: 'bundle',
} as const;

export interface ContentKey {
  prefix: string;
  parts: string[];
  key: string;
}

export function policyManifestKey(policyId: string, version: number): ContentKey {
  const parts = [KEY_PREFIX.POLICY_MANIFEST, policyId, 'manifest', String(version)];
  return { prefix: KEY_PREFIX.POLICY_MANIFEST, parts, key: parts.join(':') };
}

export function manifestDigestKey(digest: string): ContentKey {
  const parts = [KEY_PREFIX.MANIFEST_DIGEST, digest];
  return { prefix: KEY_PREFIX.MANIFEST_DIGEST, parts, key: parts.join(':') };
}

export function scriptKey(scriptHash: string): ContentKey {
  const parts = [KEY_PREFIX.SCRIPT, scriptHash];
  return { prefix: KEY_PREFIX.SCRIPT, parts, key: parts.join(':') };
}

export function proofKey(policyRoot: string, scriptHash: string): ContentKey {
  const parts = [KEY_PREFIX.PROOF, policyRoot, scriptHash];
  return { prefix: KEY_PREFIX.PROOF, parts, key: parts.join(':') };
}

export function bundleKey(bundleHash: string): ContentKey {
  const parts = [KEY_PREFIX.BUNDLE, bundleHash];
  return { prefix: KEY_PREFIX.BUNDLE, parts, key: parts.join(':') };
}

export function parseContentKey(key: string): ContentKey | null {
  const parts = key.split(':');
  if (parts.length < 2) return null;
  const prefix = parts[0];
  if (!Object.values(KEY_PREFIX).includes(prefix as typeof KEY_PREFIX[keyof typeof KEY_PREFIX])) {
    return null;
  }
  return { prefix, parts, key };
}

export function computeManifestDigest(manifestBytes: Uint8Array): string {
  return bytesToHex(sha3_256(manifestBytes));
}

export function computeBundleHash(manifest: Uint8Array, branches: Uint8Array[]): string {
  const parts: Uint8Array[] = [manifest];
  for (const b of branches) parts.push(b);
  const totalLen = parts.reduce((s, p) => s + p.length, 0);
  const combined = new Uint8Array(totalLen);
  let offset = 0;
  for (const p of parts) {
    combined.set(p, offset);
    offset += p.length;
  }
  return bytesToHex(sha3_256(combined));
}

export function computeScriptHash(script: string): string {
  return bytesToHex(sha3_256(new TextEncoder().encode(script)));
}

File: packages/totem-sdk/packages/recursive-mast/rust/src/content_keys.rs
use sha3::{Digest, Sha3_256};

use crate::types::ContentKey;

pub const KEY_PREFIX_POLICY_MANIFEST: &str = "policy";
pub const KEY_PREFIX_MANIFEST_DIGEST: &str = "manifest";
pub const KEY_PREFIX_SCRIPT: &str = "script";
pub const KEY_PREFIX_PROOF: &str = "proof";
pub const KEY_PREFIX_BUNDLE: &str = "bundle";

pub fn policy_manifest_key(policy_id: &str, version: u32) -> ContentKey {
    let parts = vec![
        KEY_PREFIX_POLICY_MANIFEST.to_string(),
        policy_id.to_string(),
        "manifest".to_string(),
        version.to_string(),
    ];
    ContentKey {
        prefix: KEY_PREFIX_POLICY_MANIFEST.to_string(),
        key: parts.join(":"),
        parts,
    }
}

pub fn manifest_digest_key(digest: &str) -> ContentKey {
    let parts = vec![
        KEY_PREFIX_MANIFEST_DIGEST.to_string(),
        digest.to_string(),
    ];
    ContentKey {
        prefix: KEY_PREFIX_MANIFEST_DIGEST.to_string(),
        key: parts.join(":"),
        parts,
    }
}

pub fn script_key(script_hash: &str) -> ContentKey {
    let parts = vec![KEY_PREFIX_SCRIPT.to_string(), script_hash.to_string()];
    ContentKey {
        prefix: KEY_PREFIX_SCRIPT.to_string(),
        key: parts.join(":"),
        parts,
    }
}

pub fn proof_key(policy_root: &str, script_hash: &str) -> ContentKey {
    let parts = vec![
        KEY_PREFIX_PROOF.to_string(),
        policy_root.to_string(),
        script_hash.to_string(),
    ];
    ContentKey {
        prefix: KEY_PREFIX_PROOF.to_string(),
        key: parts.join(":"),
        parts,
    }
}

pub fn bundle_key(bundle_hash: &str) -> ContentKey {
    let parts = vec![KEY_PREFIX_BUNDLE.to_string(), bundle_hash.to_string()];
    ContentKey {
        prefix: KEY_PREFIX_BUNDLE.to_string(),
        key: parts.join(":"),
        parts,
    }
}

pub fn parse_content_key(key: &str) -> Option<ContentKey> {
    let parts: Vec<String> = key.split(':').map(|s| s.to_string()).collect();
    if parts.len() < 2 {
        return None;
    }
    let valid_prefixes = [
        KEY_PREFIX_POLICY_MANIFEST,
        KEY_PREFIX_MANIFEST_DIGEST,
        KEY_PREFIX_SCRIPT,
        KEY_PREFIX_PROOF,
        KEY_PREFIX_BUNDLE,
    ];
    if !valid_prefixes.contains(&parts[0].as_str()) {
        return None;
    }
    Some(ContentKey {
        prefix: parts[0].clone(),
        key: key.to_string(),
        parts,
    })
}

pub fn compute_manifest_digest(manifest_bytes: &[u8]) -> String {
    let mut hasher = Sha3_256::new();
    hasher.update(manifest_bytes);
    hex::encode(hasher.finalize())
}

pub fn compute_bundle_hash(manifest: &[u8], branches: &[Vec<u8>]) -> String {
    let total_len: usize = manifest.len() + branches.iter().map(|b| b.len()).sum::<usize>();
    let mut combined = Vec::with_capacity(total_len);
    combined.extend_from_slice(manifest);
    for branch in branches {
        combined.extend_from_slice(branch);
    }
    let mut hasher = Sha3_256::new();
    hasher.update(&combined);
    hex::encode(hasher.finalize())
}

pub fn compute_script_hash(script: &str) -> String {
    let mut hasher = Sha3_256::new();
    hasher.update(script.as_bytes());
    hex::encode(hasher.finalize())
}
